harpoon dynamic commands for projects

RishabhRD opened the pull request:
As discussed in #96 , this PR adds dynamic_commands feature to harpoon.

I have done some additions in harpoon.term module to provide API to add and remove commands.
I have added another module named harpoon.cmd-ui, this module intends to provide quick menu features like marks quick menu.

Sample usecase video link: https://youtu.be/oI4kMpkADI4

ThePrimeagen commented on lua/harpoon/cmd-ui.lua:
```diff
@@ -0,0 +1,148 @@
+end
+
+local function is_white_space(str)
+    local white_space_chars = {
```
we could probably just use \s and a regex here.  I KNOW... I SAID THAT... 

There should be a gsub call you can do, something like:

```
return #(str:gsub('%s', '')) == 0 -- not sure the exact api
```

RishabhRD replied:
done the change

ThePrimeagen commented on lua/harpoon/cmd-ui.lua:
```diff
@@ -0,0 +1,148 @@
+end
+
+M.toggle_quick_menu = function()
+    log.trace("toggle_quick_menu()")
```
add cmd-ui to the trace so we know which toggle quick menu.

ThePrimeagen commented on lua/harpoon/cmd-ui.lua:
```diff
@@ -0,0 +1,148 @@
+
+local M = {}
+
+Harpoon_cmd_win_id = nil
```
I know I wrote this in the past, but I hate these global vars.  I feel like we shouldn't be using them :)

But we are currently, so I'll let it slide, but if I could get my way it would be to take these pop-up windows and just turn it into a class.

ThePrimeagen commented on lua/harpoon/cmd-ui.lua:
```diff
@@ -0,0 +1,148 @@
+end
+
+M.on_menu_save = function()
+    log.trace("on_menu_save()")
```
cmd-ui#on_menu_save()

ThePrimeagen commented on lua/harpoon/cmd-ui.lua:
```diff
@@ -0,0 +1,148 @@
+end
+
+local function create_window()
+    log.trace("_create_window()")
```
I am a bit worried about so much duplicated logic.  I just don't want to create bugs that are hard to squuuuuoooosh

ThePrimeagen reviewed (approved):
I am going to merge this PR but I really do worry about the copied logic. 

I'll have to strongly consider reducing the two items used here, pop up windows + autocommands and "holed" lists into a class.  That way mutations can emit events easily without having to have this trixy logic spread around.

ThePrimeagen commented:
The only thing I ask is that you make the whitespace more simple.  For this case, the regex just works better since it will encompass all white space.  In general, the white space checking with regex's seems to be one of the only use cases I use for them.

RishabhRD commented:
@ThePrimeagen Done with whitespace change and logger breadbutter. Didn't change the traces for private local functions because they would be called in context of one of public functions.

I can do the removal of code duplication by making classes for window and some free generic functions for holed lists... but that seems to be a refactor and would change the other files in project.... that doesn't seem me suitable for this PR. Am I right?

ThePrimeagen commented:
I'll merge this, try it out, and hopefully its awesome :)

ThePrimeagen commented:
@RishabhRD you have some lint :)

RishabhRD commented:
@ThePrimeagen I have updated luacheckrc for ignoring the global variables I have introduced as other global variables are also there as those would be refactored out in future versions. Also done some formatting things as formatting was failing.

RishabhRD commented:
I need to change formatting I guess. Actually formatting is quite messy with stylua. I don't know I tried to format document 3 times with given .stylua.toml but it said no changes required :sweat_smile: .

RishabhRD commented:
@ThePrimeagen I fixed the formatting and tested it using stylua also checked lint using luacheck... Please reivew it once

RishabhRD commented:
@ThePrimeagen any update on this PR?

ThePrimeagen commented:
Sorry this front end Masters presentation has been eating up all my time. Let me review this as soon as possible.

ThePrimeagen commented:
Uniroically I have merged this without review! I just want to make you happy! Will review later ^^

RishabhRD commented:
Thanks... I hope it would be useful for many :)
